File: src/main/javascript-mvc-1.5/jmvc/plugins/reflection/reflection_utils.js
/*
 * The main contributor to this project is Institute of Materials Research,
 * Helmholtz-Zentrum Geesthacht,
 * Germany.
 *
 * This project is a contribution of the Helmholtz Association Centres and
 * Technische Universitaet Muenchen to the ESS Design Update Phase.
 *
 * The project's funding reference is FKZ05E11CG1.
 *
 * Copyright (c) 2012. Institute of Materials Research,
 * Helmholtz-Zentrum Geesthacht,
 * Germany.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 */

/**
 * Created by
 * User: khokhria
 * Date: 18.01.12
 */
MVC.ReflectionUtils = (function ($) {
    return {
        /**
         * Creates a new Class instance. New instance extends MVC.Model
         *
         * @param className a class name
         * @param attributes must contain 'domain' and 'meta' values
         * @param properties prototype properties
         */
        createModel:function (className, attributes, properties) {

            var clazz = MVC.Model./*JsonP.*/extend(className, {
                domain:attributes.domain,
                attributes:{},
                default_attributes:{},
                init:function () {
                    var me = this;
                    $.each(attributes.meta.fields, function (i, field) {
                        //TODO type based factory
                        if (field.type == 'choice') {
                            me.add_attribute(field.id, 'boolean');
                            $.each(field.fields, function (i, field) {
                                me.add_attribute(field.id, field.type);
                            })
                        } else {
                            me.add_attribute(field.id, field.type);
                        }
                    });
                    me._super();
                },
                update:function (instance/*void*/, params, callbacks) {
                    this.create(params, callbacks);
                }
            }, {

            });
            window[className] = clazz;
            return clazz;
        }
    }
})(jQuery);

File: src/main/java/wpn/hdri/util/reflection/ReflectionUtils.java
/*
 * The main contributor to this project is Institute of Materials Research,
 * Helmholtz-Zentrum Geesthacht,
 * Germany.
 *
 * This project is a contribution of the Helmholtz Association Centres and
 * Technische Universitaet Muenchen to the ESS Design Update Phase.
 *
 * The project's funding reference is FKZ05E11CG1.
 *
 * Copyright (c) 2012. Institute of Materials Research,
 * Helmholtz-Zentrum Geesthacht,
 * Germany.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 */

package wpn.hdri.util.reflection;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Igor Khokhriakov <igor.khokhriakov@hzg.de>
 * @since 28.02.12
 */
public class ReflectionUtils {
    private ReflectionUtils() {
    }

    /**
     * Acts as {@link Class#getDeclaredFields()} but also includes inherited fields.
     *
     * @param clazz
     * @return
     */
    public static Field[] getDeclaredFields(Class<?> clazz) {
        List<Field> fields = new ArrayList<Field>();

        do {
            Field[] dclFields = clazz.getDeclaredFields();
            fields.addAll(Arrays.asList(dclFields));
        } while ((clazz = clazz.getSuperclass()) != null);

        return fields.toArray(new Field[fields.size()]);
    }

    /**
     * Acts as {@link Class#getDeclaredField(String)} but also looks in super classes for the field.
     *
     * @param name  a name of the field to look up
     * @param clazz a class
     * @return Field or throws exception
     * @throws Exception if no field with the given name was found in the class hierarchy
     */
    public static Field getDeclaredField(String name, Class<?> clazz) throws Exception {
        do {
            try {
                return clazz.getDeclaredField(name);
            } catch (NoSuchFieldException ignored) {
            } catch (SecurityException e) {
                throw new Exception(e);
            }
        } while ((clazz = clazz.getSuperclass()) != null);
        throw new NoSuchFieldException(name);
    }
}
